File: gcj_2018/QR_A.hs
-- Google Code Jam 2018 Qualification Round
-- Problem A. Saving The Universe Again

import Control.Monad
import Data.Char
import Data.List
import Text.Printf

readInt = (read <$> getLine) :: IO Int

findSubstring :: Eq a => [a] -> [a] -> Maybe Int
findSubstring pat str = findIndex (isPrefixOf pat) (tails str)

total :: String -> Int -> Int -> Int
total s sum power
  | (length s) == 0 = sum
  | (head s) == 'C' = total (tail s) sum (2 * power)
  | (head s) == 'S' = total (tail s) (sum + power) power
  | otherwise = 0

f :: Int -> String -> Int -> Int
f d s t = do
  if (total (reverse s) 0 1) <= d then t else do
    case findSubstring "SC" s of
      Just x -> f d (((take x s) ++ "CS") ++ (drop (x + 2) s)) (t + 1)
      _ -> -1

solve :: Int -> String -> String
solve d s = do
  let ans = f d (reverse s) 0
  if ans < 0 then "IMPOSSIBLE" else show ans

main :: IO ()
main = do
  t <- readInt
  forM_ [1..t] $ \i -> do
    [d, s] <- words <$> getLine
    putStrLn $ "Case #" ++ (show i) ++ ": " ++ (solve (read d :: Int) s)

File: gcj_2018/QR_A.cpp
// Google Code Jam 2018 Qualification Round
// Problem A. Saving The Universe Again

#include <iostream>
#include <sstream>

using namespace std;

typedef long long LL;

int solve(LL d, string s) {
	for (int t = 0; ; ++t) {
			LL sum = 0;
		LL power = 1;
		for (char c : s) {
			if (c == 'C') {
				power *= 2;
			}
			if (c == 'S') {
				sum += power;
			}
		}
		if (sum <= d) {
			return t;
		}
		int pos = (int)s.rfind("CS");
		if (pos < 0) {
			break;
		}
		s = s.substr(0, pos) + "SC" + s.substr(pos + 2, s.length() - pos - 2);
	}
	return -1;
}

int main(int argc, char *argv[]) {
	int T;
	cin >> T;
	for (int t = 1; t <= T; ++t) {
		LL d;
		string s;
		cin >> d >> s;
		int ans = solve(d, s);
		if (ans < 0) {
			cout << "Case #" << t << ": IMPOSSIBLE" << endl;
		} else {
			cout << "Case #" << t << ": " << ans << endl;
		}
	}
	return 0;
}
